Add possibility to add custom annotations to Builder

In particular `edu.umd.cs.findbugs.annotations.SuppressFBWarnings`, otherweise [SpotBugs](https://spotbugs.github.io/) detects violations which normally would be ignored in generated code.

It's not possible to make SpotBugs process `@Generated` annotation since SpotBugs process byte-code and `@Generated` has `source` retention type.

P.S. Nice library and nice work! I use [Immutables](https://immutables.github.io/) a lot so it's really nice to have similar funtionality for java records.

For these annotations specifically, IDEs will warn developers if they pass `null` to a setter where the parameter is annotated with `@Nonnull`. So with the example above, if the annotations were copied I would get the following from my IDE (IntelliJ IDEA):
```java
Address.builder()
        .country(null) // <- Warning: Passing 'null' argument to parameter annotated as @NotNull
        .build()
```
This can be VERY useful to identify whether you're using a record as it is intended. It can also be reported during the build process (and potentially force a build to fail) with plugins such as [SpotBugs](https://spotbugs.github.io/):
```
[INFO] --- spotbugs-maven-plugin:4.2.0:check (default-cli) @ my-project ---
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
[ERROR] High: Null passed for non-null parameter of my.package.Address$AddressBuilder.country(String) in my.package.App.createAddress() [my.package.App] At App.java:[line 54] NP_NONNULL_PARAM_VIOLATION
```

--- a/pom.xml
+++ b/pom.xml
@@ -37,6 +37,7 @@
         <javapoet-version>1.12.1</javapoet-version>
         <junit-jupiter-version>5.5.2</junit-jupiter-version>
         <asm-version>7.2</asm-version>
+        <validation-api-version>2.0.1.Final</validation-api-version>
     </properties>
 
     <name>Record Builder</name>
@@ -112,6 +113,12 @@
                 <groupId>org.junit.jupiter</groupId>
                 <artifactId>junit-jupiter</artifactId>
                 <version>${junit-jupiter-version}</version>
+            </dependency>
+
+            <dependency>
+                <groupId>javax.validation</groupId>
+                <artifactId>validation-api</artifactId>
+                <version>${validation-api-version}</version>
             </dependency>
         </dependencies>
     </dependencyManagement>

--- a/record-builder-core/src/main/java/io/soabase/recordbuilder/core/RecordBuilder.java
+++ b/record-builder-core/src/main/java/io/soabase/recordbuilder/core/RecordBuilder.java
@@ -108,6 +108,14 @@ public @interface RecordBuilder {
          * be prefixed to the builder name if this returns true.
          */
         boolean prefixEnclosingClassNames() default true;
+
+        /**
+         * If true, any annotations (if applicable) on record components are copied
+         * to the builder methods
+         *
+         * @return true/false
+         */
+        boolean inheritComponentAnnotations() default true;
     }
 
     @Retention(RetentionPolicy.SOURCE)

--- a/record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/ElementUtils.java
+++ b/record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/ElementUtils.java
@@ -24,6 +24,7 @@ import javax.annotation.processing.ProcessingEnvironment;
 import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.AnnotationValue;
 import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.element.RecordComponentElement;
 import javax.lang.model.element.TypeElement;
@@ -107,8 +108,8 @@ public class ElementUtils {
         return new ClassType(ParameterizedTypeName.get(builderClassName, typeNames), builderClassName.simpleName());
     }
 
-    public static ClassType getClassType(RecordComponentElement recordComponent) {
-        return new ClassType(TypeName.get(recordComponent.asType()), recordComponent.getSimpleName().toString());
+    public static RecordClassType getRecordClassType(RecordComponentElement recordComponent, List<? extends AnnotationMirror> accessorAnnotations, List<? extends AnnotationMirror> canonicalConstructorAnnotations) {
+        return new RecordClassType(TypeName.get(recordComponent.asType()), recordComponent.getSimpleName().toString(), accessorAnnotations, canonicalConstructorAnnotations);
     }
 
     public static String getWithMethodName(ClassType component, String prefix) {
@@ -123,6 +124,23 @@ public class ElementUtils {
         // generate the class name
         var baseName = classType.name() + suffix;
         return metaData.prefixEnclosingClassNames() ? (getBuilderNamePrefix(element.getEnclosingElement()) + baseName) : baseName;
+    }
+
+    public static Optional<? extends Element> findCanonicalConstructor(TypeElement record) {
+        if ( record.getKind() != ElementKind.RECORD ) {
+            return Optional.empty();
+        }
+
+        // based on https://github.com/openjdk/jdk/pull/3556/files#diff-a6270f4b50989abe733607c69038b2036306d13f77276af005d023b7fc57f1a2R2368
+        var componentList = record.getRecordComponents().stream().map(e -> e.asType().toString()).collect(Collectors.toList());
+        return record.getEnclosedElements().stream()
+            .filter(element -> element.getKind() == ElementKind.CONSTRUCTOR)
+            .filter(element -> {
+                var parameters = ((ExecutableElement)element).getParameters();
+                var parametersList = parameters.stream().map(e -> e.asType().toString()).collect(Collectors.toList());
+                return componentList.equals(parametersList);
+            })
+            .findFirst();
     }
 
     private static String getBuilderNamePrefix(Element element) {

--- a/record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/InternalRecordBuilderProcessor.java
+++ b/record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/InternalRecordBuilderProcessor.java
@@ -15,6 +15,7 @@
  */
 package io.soabase.recordbuilder.processor;
 
+import com.squareup.javapoet.AnnotationSpec;
 import com.squareup.javapoet.ClassName;
 import com.squareup.javapoet.CodeBlock;
 import com.squareup.javapoet.FieldSpec;
@@ -25,8 +26,9 @@ import com.squareup.javapoet.TypeName;
 import com.squareup.javapoet.TypeSpec;
 import com.squareup.javapoet.TypeVariableName;
 import io.soabase.recordbuilder.core.RecordBuilder;
-import javax.lang.model.element.Modifier;
-import javax.lang.model.element.TypeElement;
+
+import javax.lang.model.element.*;
+
 import java.util.AbstractMap;
 import java.util.List;
 import java.util.Map;
@@ -48,7 +50,7 @@ class InternalRecordBuilderProcessor
     private final String packageName;
     private final ClassType builderClassType;
     private final List<TypeVariableName> typeVariables;
-    private final List<ClassType> recordComponents;
+    private final List<RecordClassType> recordComponents;
     private final TypeSpec builderType;
     private final TypeSpec.Builder builder;
     private final String uniqueVarName;
@@ -60,7 +62,7 @@ class InternalRecordBuilderProcessor
         packageName = packageNameOpt.orElseGet(() -> ElementUtils.getPackageName(record));
         builderClassType = ElementUtils.getClassType(packageName, getBuilderName(record, metaData, recordClassType, metaData.suffix()), record.getTypeParameters());
         typeVariables = record.getTypeParameters().stream().map(TypeVariableName::get).collect(Collectors.toList());
-        recordComponents = record.getRecordComponents().stream().map(ElementUtils::getClassType).collect(Collectors.toList());
+        recordComponents = buildRecordComponents(record);
         uniqueVarName = getUniqueVarName();
 
         builder = TypeSpec.classBuilder(builderClassType.name())
@@ -104,9 +106,24 @@ class InternalRecordBuilderProcessor
         return builderType;
     }
 
-    private RecordBuilder.Options getMetaData(TypeElement record, RecordBuilder.Options metaData) {
+    private List<RecordClassType> buildRecordComponents(TypeElement record)
+    {
+        var accessorAnnotations = record.getRecordComponents().stream().map(e -> e.getAccessor().getAnnotationMirrors()).collect(Collectors.toList());
+        var canonicalConstructorAnnotations = ElementUtils.findCanonicalConstructor(record).map(constructor -> ((ExecutableElement) constructor).getParameters().stream().map(Element::getAnnotationMirrors).collect(Collectors.toList())).orElse(List.of());
+        var recordComponents = record.getRecordComponents();
+        return IntStream.range(0, recordComponents.size())
+                .mapToObj(index -> {
+                    var thisAccessorAnnotations = (accessorAnnotations.size() > index) ? accessorAnnotations.get(index) : List.<AnnotationMirror>of();
+                    var thisCanonicalConstructorAnnotations = (canonicalConstructorAnnotations.size() > index) ? canonicalConstructorAnnotations.get(index) : List.<AnnotationMirror>of();
+                    return ElementUtils.getRecordClassType(recordComponents.get(index), thisAccessorAnnotations, thisCanonicalConstructorAnnotations);
+                })
+                .collect(Collectors.toList());
+    }
+
+    private RecordBuilder.Options getMetaData(TypeElement record, RecordBuilder.Options metaData)
+    {
         var recordSpecificMetaData = record.getAnnotation(RecordBuilder.Options.class);
-        return (recordSpecificMetaData != null) ?recordSpecificMetaData : metaData;
+        return (recordSpecificMetaData != null) ? recordSpecificMetaData : metaData;
     }
 
     private void addWithNestedClass()
@@ -191,12 +208,12 @@ class InternalRecordBuilderProcessor
     {
         var name = prefix + "r";
         var alreadyExists = recordComponents.stream()
-            .map(ClassType::name)
-            .anyMatch(n -> n.equals(name));
+                .map(ClassType::name)
+                .anyMatch(n -> n.equals(name));
         return alreadyExists ? getUniqueVarName(prefix + "_") : name;
     }
 
-    private void add1WithMethod(TypeSpec.Builder classBuilder, ClassType component, int index)
+    private void add1WithMethod(TypeSpec.Builder classBuilder, RecordClassType component, int index)
     {
         /*
             Adds a with method for the component similar to:
@@ -226,12 +243,13 @@ class InternalRecordBuilderProcessor
         codeBlockBuilder.add(");");
 
         var methodName = getWithMethodName(component, metaData.withClassMethodPrefix());
-        var parameterSpec = ParameterSpec.builder(component.typeName(), component.name()).build();
+        var parameterSpecBuilder = ParameterSpec.builder(component.typeName(), component.name());
+        component.getCanonicalConstructorAnnotations().forEach(annotationMirror -> parameterSpecBuilder.addAnnotation(AnnotationSpec.get(annotationMirror)));
         var methodSpec = MethodSpec.methodBuilder(methodName)
                 .addAnnotation(generatedRecordBuilderAnnotation)
                 .addJavadoc("Return a new instance of {@code $L} with a new value for {@code $L}\n", recordClassType.name(), component.name())
                 .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
-                .addParameter(parameterSpec)
+                .addParameter(parameterSpecBuilder.build())
                 .addCode(codeBlockBuilder.build())
                 .returns(recordClassType.typeName())
                 .build();
@@ -270,7 +288,11 @@ class InternalRecordBuilderProcessor
                 .addAnnotation(generatedRecordBuilderAnnotation)
                 .returns(recordClassType.typeName())
                 .addStatement(codeBlock);
-        recordComponents.forEach(component -> builder.addParameter(component.typeName(), component.name()));
+        recordComponents.forEach(component -> {
+            var parameterSpecBuilder = ParameterSpec.builder(component.typeName(), component.name());
+            component.getCanonicalConstructorAnnotations().forEach(annotationMirror -> parameterSpecBuilder.addAnnotation(AnnotationSpec.get(annotationMirror)));
+            builder.addParameter(parameterSpecBuilder.build());
+        });
         this.builder.addMethod(builder.build());
     }
 
@@ -412,7 +434,8 @@ class InternalRecordBuilderProcessor
         builder.addMethod(methodSpec);
     }
 
-    private CodeBlock buildCodeBlock() {
+    private CodeBlock buildCodeBlock()
+    {
         /*
             Builds the code block for allocating the record from its parts
         */
@@ -523,25 +546,25 @@ class InternalRecordBuilderProcessor
             }
          */
         var codeBlockBuilder = CodeBlock.builder()
-            .add("try {\n")
-            .indent()
-            .add("return ($T)obj;\n", recordClassType.typeName())
-            .unindent()
-            .add("}\n")
-            .add("catch (ClassCastException dummy) {\n")
-            .indent()
-            .add("throw new RuntimeException($S);\n", builderClassType.name() + "." + metaData.withClassName() + " can only be implemented by " + recordClassType.name())
-            .unindent()
-            .add("}");
+                .add("try {\n")
+                .indent()
+                .add("return ($T)obj;\n", recordClassType.typeName())
+                .unindent()
+                .add("}\n")
+                .add("catch (ClassCastException dummy) {\n")
+                .indent()
+                .add("throw new RuntimeException($S);\n", builderClassType.name() + "." + metaData.withClassName() + " can only be implemented by " + recordClassType.name())
+                .unindent()
+                .add("}");
         var methodSpec = MethodSpec.methodBuilder(metaData.downCastMethodName())
-            .addAnnotation(generatedRecordBuilderAnnotation)
-            .addJavadoc("Downcast to {@code $L}\n", recordClassType.name())
-            .addModifiers(Modifier.PRIVATE, Modifier.STATIC)
-            .addParameter(Object.class, "obj")
-            .addTypeVariables(typeVariables)
-            .returns(recordClassType.typeName())
-            .addCode(codeBlockBuilder.build())
-            .build();
+                .addAnnotation(generatedRecordBuilderAnnotation)
+                .addJavadoc("Downcast to {@code $L}\n", recordClassType.name())
+                .addModifiers(Modifier.PRIVATE, Modifier.STATIC)
+                .addParameter(Object.class, "obj")
+                .addTypeVariables(typeVariables)
+                .returns(recordClassType.typeName())
+                .addCode(codeBlockBuilder.build())
+                .build();
         builder.addMethod(methodSpec);
     }
 
@@ -556,7 +579,7 @@ class InternalRecordBuilderProcessor
         builder.addField(fieldSpec);
     }
 
-    private void add1GetterMethod(ClassType component)
+    private void add1GetterMethod(RecordClassType component)
     {
         /*
             For a single record component, add a getter similar to:
@@ -565,17 +588,17 @@ class InternalRecordBuilderProcessor
                 return p;
             }
          */
-        var methodSpec = MethodSpec.methodBuilder(component.name())
+        var methodSpecBuilder = MethodSpec.methodBuilder(component.name())
                 .addJavadoc("Return the current value for the {@code $L} record component in the builder\n", component.name())
                 .addModifiers(Modifier.PUBLIC)
                 .addAnnotation(generatedRecordBuilderAnnotation)
                 .returns(component.typeName())
-                .addStatement("return $L", component.name())
-                .build();
-        builder.addMethod(methodSpec);
+                .addStatement("return $L", component.name());
+        component.getAccessorAnnotations().forEach(annotationMirror -> methodSpecBuilder.addAnnotation(AnnotationSpec.get(annotationMirror)));
+        builder.addMethod(methodSpecBuilder.build());
     }
 
-    private void add1SetterMethod(ClassType component)
+    private void add1SetterMethod(RecordClassType component)
     {
         /*
             For a single record component, add a setter similar to:
@@ -585,12 +608,14 @@ class InternalRecordBuilderProcessor
                 return this;
             }
          */
-        var parameterSpec = ParameterSpec.builder(component.typeName(), component.name()).build();
+        var parameterSpecBuilder = ParameterSpec.builder(component.typeName(), component.name());
+        component.getCanonicalConstructorAnnotations().forEach(annotationMirror -> parameterSpecBuilder.addAnnotation(AnnotationSpec.get(annotationMirror)));
+
         var methodSpec = MethodSpec.methodBuilder(component.name())
                 .addJavadoc("Set a new value for the {@code $L} record component in the builder\n", component.name())
                 .addModifiers(Modifier.PUBLIC)
                 .addAnnotation(generatedRecordBuilderAnnotation)
-                .addParameter(parameterSpec)
+                .addParameter(parameterSpecBuilder.build())
                 .returns(builderClassType.typeName())
                 .addStatement("this.$L = $L", component.name(), component.name())
                 .addStatement("return this")

--- a/record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/RecordClassType.java
+++ b/record-builder-processor/src/main/java/io/soabase/recordbuilder/processor/RecordClassType.java
@@ -1,0 +1,40 @@
+/**
+ * Copyright 2019 Jordan Zimmerman
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.soabase.recordbuilder.processor;
+
+import com.squareup.javapoet.TypeName;
+
+import javax.lang.model.element.AnnotationMirror;
+import java.util.List;
+
+public class RecordClassType extends ClassType {
+    private final List<? extends AnnotationMirror> accessorAnnotations;
+    private final List<? extends AnnotationMirror> canonicalConstructorAnnotations;
+
+    public RecordClassType(TypeName typeName, String name, List<? extends AnnotationMirror> accessorAnnotations, List<? extends AnnotationMirror> canonicalConstructorAnnotations) {
+        super(typeName, name);
+        this.accessorAnnotations = accessorAnnotations;
+        this.canonicalConstructorAnnotations = canonicalConstructorAnnotations;
+    }
+
+    public List<? extends AnnotationMirror> getAccessorAnnotations() {
+        return accessorAnnotations;
+    }
+
+    public List<? extends AnnotationMirror> getCanonicalConstructorAnnotations() {
+        return canonicalConstructorAnnotations;
+    }
+}

--- a/record-builder-test/pom.xml
+++ b/record-builder-test/pom.xml
@@ -11,6 +11,11 @@
 
     <dependencies>
         <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+        </dependency>
+
+        <dependency>
             <groupId>io.soabase.record-builder</groupId>
             <artifactId>record-builder-core</artifactId>
             <scope>provided</scope>

--- a/record-builder-test/src/main/java/io/soabase/recordbuilder/test/Annotated.java
+++ b/record-builder-test/src/main/java/io/soabase/recordbuilder/test/Annotated.java
@@ -1,0 +1,26 @@
+/**
+ * Copyright 2019 Jordan Zimmerman
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.soabase.recordbuilder.test;
+
+import io.soabase.recordbuilder.core.RecordBuilder;
+import javax.validation.constraints.Max;
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Null;
+
+@RecordBuilder
+public record Annotated(@NotNull @Null String hey, @Min(10) @Max(100) int i, double d) {
+}

